File: app/src/main/java/com/dot2line/vocabar/base/BaseMVPActivity.java
package com.dot2line.vocabar.base;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

public abstract class BaseMVPActivity<V extends BaseMVPView, P extends BaseMVPPresenter<V>> extends AppCompatActivity {
    protected V baseView;
    protected P basePresenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        basePresenter = createPresenter();
        baseView = createView();
    }

    @Override
    protected void onStart() {
        super.onStart();

        basePresenter.takeView(baseView);
    }

    @Override
    protected void onStop() {
        super.onStop();

        basePresenter.dropView(baseView);
    }

    protected abstract P createPresenter();

    protected abstract V createView();
}

File: app/src/main/java/com/dot2line/vocabar/base/BaseMVPActivity.kt
package com.dot2line.vocabar.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

abstract class BaseMVPActivity<V : BaseMVPView, P : BaseMVPPresenter<V>> : AppCompatActivity() {
  lateinit var baseView: V
  lateinit var basePresenter: P

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    basePresenter = createPresenter()
    baseView = createView()
  }

  override fun onStart() {
    super.onStart()

    basePresenter.takeView(baseView)
  }

  override fun onStop() {
    super.onStop()

    basePresenter.dropView(baseView)
  }

  protected abstract fun createPresenter(): P

  protected abstract fun createView(): V
}